pkg/util/tracing/collector/collector.go

```diff
@@ -53,59 +50,125 @@ func New(
-		alloc, err := qp.Acquire(ctx, 1)
-		if err != nil {
-			return res, err
+	// Reset buffer variables.
+	i.recordedSpans = nil
+	i.recordedSpanIndex = 0
+
+	// Otherwise, either there are no more spans or we have exhausted the
+	// recordings from the current node, and we need to pull the inflight
+	// recordings from another node.
+	// Keep searching for recordings from all live nodes in the cluster.
+	for i.recordedSpans == nil {
+		i.curNodeIndex++
+		// No more spans to return from any of the live nodes in the cluster.
+		if !(i.curNodeIndex < len(i.liveNodes)) {
+			return false
+		}
+		i.recordedSpans, i.iterErr = i.collector.getTraceSpanRecordingsForNode(i.ctx, i.traceID,
+			i.liveNodes[i.curNodeIndex])
+		// TODO(adityamaru): We might want to consider not failing if a single node
+		// fails to return span recordings.
+		if i.iterErr != nil {
+			return false
 		}
+	}
+
+	return true
+}
+
+// Next sets the iterator index to point to the next buffered RecordedSpan.
+func (i *TraceCollectorIterator) Next() {
+	i.recordedSpanIndex++
+}
+
+// Value returns the RecordedSpan pointed to by the iterator.
+func (i *TraceCollectorIterator) Value() tracingpb.RecordedSpan {
+	return i.recordedSpans[i.recordedSpanIndex]
+}
+
+// Error returns the error encountered by the iterator when fetching inflight
+// span recordings.
+func (i *TraceCollectorIterator) Error() error {
+	return i.iterErr
+}
 
-		grp.GoCtx(func(ctx context.Context) error {
-			defer alloc.Release()
-
-			conn, err := t.dialer.Dial(ctx, id, rpc.DefaultClass)
-			if err != nil {
-				return err
-			}
-			traceClient := tracingservicepb.NewTracingClient(conn)
-			resp, err := traceClient.GetSpanRecordings(ctx,
-				&tracingservicepb.SpanRecordingRequest{TraceID: traceID})
-			if err != nil {
-				return err
-			}
-			mu.Lock()
-			res = append(res, resp.SpanRecordings...)
-			mu.Unlock()
-			return nil
-		})
+// getTraceSpanRecordingsForNode returns the inflight span recordings for traces
+// with traceID from the node with nodeID. The span recordings are sorted by
+// StartTime.
+// This method does not distinguish between requests for local and remote
+// inflight spans, and relies on gRPC short circuiting local requests.
+func (t *TraceCollector) getTraceSpanRecordingsForNode(
+	ctx context.Context, traceID uint64, nodeID roachpb.NodeID,
+) ([]tracingpb.RecordedSpan, error) {
+	log.Infof(ctx, "executing GetSpanRecordings on nodes %s", nodeID.String())
+	conn, err := t.dialer.Dial(ctx, nodeID, rpc.DefaultClass)
+	if err != nil {
+		return nil, err
 	}
-	if err := grp.Wait(); err != nil {
-		return res, err
+	traceClient := tracingservicepb.NewTracingClient(conn)
+	resp, err := traceClient.GetSpanRecordings(ctx,
+		&tracingservicepb.SpanRecordingRequest{TraceID: traceID})
+	if err != nil {
+		return nil, err
 	}
 
-	sort.SliceStable(res, func(i, j int) bool {
-		return res[i].StartTime.Before(res[j].StartTime)
+	sort.SliceStable(resp.SpanRecordings, func(i, j int) bool {
```

Comment:
I'm sure you're aware, but one thing we're giving up by iter-ifying this retrieval is the global sort order of all recordings. I assume that's fine for what you want. Does it make sense to avoid sorting it here anyway? If the caller needs a global sort, they should do it themselves (do we need it in our tests to avoid flakiness?).

Reply:
I think I'd like to keep the per-node sorting. I foresee the overlying tool writing span recordings from a particular node in its own file. Once I have a better idea of that I can revisit if a per node sort is needed.

We already do a global sort in `getSpansFromAllNodes` in the test.